### PR TITLE
[BEAM-12544] Add support for Avro timestamps in microseconds

### PR DESCRIPTION
**Pull request by Dobiasd**
`GenericRecordToRowFn` in `AvroUtils` does not support the logical Avro type "timestamp-micros". Instead of converting it to `FieldType.DATETIME` (as it does with "timestamp-millis") it just interprets it as a raw LONG.

This pull request fixes that.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] ~Update `CHANGES.md` with noteworthy changes.~
 - [ ] ~If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).~

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java [hunk 1]
```diff
@@ -765,8 +765,8 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
       if (logicalType instanceof LogicalTypes.Decimal) {
         fieldType = FieldType.DECIMAL;
       } else if (logicalType instanceof LogicalTypes.TimestampMillis) {
-        // TODO: There is a desire to move Beam schema DATETIME to a micros representation. When
```
Comment (Dobiasd) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java:768]: This is unrelated to the representation of Beam schema `DATETIME`. The correct TODO comment for this is here: https://github.com/apache/beam/blob/961cf0936d7065bf6089e666fbbc6b83da266df2/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java#L883-L884

**Comment by Dobiasd** [comment]
R: @iemejia

**Review by iemejia — CHANGES_REQUESTED** [review]
Thanks for the contribution @Dobiasd!

Can you please add (or extend) a test that validates the `time-micros` conversion (similar to the other ones for LogicalTypes in the `AvroUtilsTest` class.

**Comment by Dobiasd** [comment]
@iemejia

Thanks for the review. :+1:

I tried to add tests (see the latest commit), but it actually is quite hard to do correctly (they currently fail), because in the Beam schema (and thus row), the Avro microsecond timestamp is also represented as `FieldType.DATETIME` (same as the millisecond timestamp). So the back conversion, i.e., `AvroUtils.toAvroSchema(beamSchema)` in `testFromBeamSchema` and `AvroUtils.toGenericRecord(getBeamRow() ...);` in `testBeamRowToGenericRecordInferSchema`/`testBeamRowToGenericRecord`, returns an Avro schema/record with milliseconds instead of microseconds, which makes the equality check with the original schema/record (`assertEquals(getAvroSchema(), avroSchema);`/`assertEquals(getGenericRecord(), genericRecord);`) fail.

Since we probably don't want to introduce a new `FieldType.DATETIMEMICROS`, we'd need to somewhat hack around this "source resolution amnesia" in the test, which might create more harm than help.

Another way would be to add an additional, separate test (without back conversion check) just for the microsecond timestamp, but regarding the triviality of the implementation, this too currently seems to me like too much overhead. What do you think?

**Comment by Dobiasd** [comment]
The tests are based on the assumption, that in the conversion chain `Avro (A) -> Beam (B) -> Avro (C)` the schema in C is the same as in A.

For millisecond timestamps this works fine: `Avro (TimestampMillis) -> Beam (DATETIME) -> Avro (TimestampMillis)`
But for microsecond timestamps it does not: `Avro (TimestampMicros) -> Beam (DATETIME) -> Avro (TimestampMillis)`

Thus I've just pushed a commit that removed the (failing) tests for `TimestampMillis` again.

@iemejia Do you think this pull request is good to (squash-)merge despite that lack of additional tests?

**Comment by aromanenko-dev** [comment]
I believe that we still need the tests if we decide that this change is valid. 

Actually, this conversion `Avro (TimestampMicros) -> Beam (DATETIME) -> Avro (TimestampMillis)` - this is really what actually is expected?

**Comment by Dobiasd** [comment]
Yeah, this is expected. It's just that the current test setup, to me at least, looks like it's not prepared for that. Any suggestions for a good approach on how to add a test?

**Comment by aromanenko-dev** [comment]
Well, in this case I'd say that we need to test it in the same way as we do for `Avro(TimestampMillis)` but expecting milliseconds after two-sides conversion which actually can be quite ambiguous from the user point perspective (and that is why I'm asking if it's expected behaviour). Then in this case we need to warn users about that in documentation as well (JavaDoc and website).

Btw, are there any principal objections or difficulties to use a micros representation for DATETIME field? 

Afaik, @iemejia is OOO this week but he will be back soon and maybe can add something on this.

**Comment by aromanenko-dev** [comment]
@Dobiasd Kind ping, any updates on this PR?

**Comment by Dobiasd** [comment]
@aromanenko-dev Sorry, not from my side. Honestly, I don't know what to do now, i.e., if I should try to understand and refactor the tests to support this new scenario, or not.

> Btw, are there any principal objections or difficulties to use a micros representation for DATETIME field?

I have no idea and I was hoping for somebody more experienced with the codebase than me to chime in here. :grimacing:

**Comment by aromanenko-dev** [comment]
cc: @RyanSkraba Could you take a look, please?

**Review by RyanSkraba — CHANGES_REQUESTED** [review]
Hello!  At this point, we're at an "opinion" about what we expect for: `Avro (A) -> Beam (B) -> Avro (C)`

I don't think "blind" round-trip identity is a strict requirement; it's fine for me that `(C)` has a different schema than `(A)`.

The Avro supports microsecond precision for dates, and the Beam row doesn't (and in fact, joda-time doesn't and [never will](https://github.com/JodaOrg/joda-time/issues/139)).

So today, if we're going to allow the A->B part, we have have a choice between 
1. retaining the date semantics and truncating the microsecords part, or
2.  keeping the full data in a Long and losing that it's a Date

I don't see a choice that would make everybody happy -- I can understand that the DateTime semantics are really important, but so are microsecond timestamps (I know that they're a legal requirement in some industries, like finance).  The Beam DATETIME would be unusable for those regulated fields, but the Beam LONG type would still work.

This PR would make the Avro `timestamp-micros` unusable in those cases without any workaround :/

Any ideas?

**Comment by aromanenko-dev** [comment]
@Dobiasd What do you think about Ryan's suggestion?

**Comment by Dobiasd** [comment]
Sorry, I feel I don't have enough understanding of the bigger picture to build/express an opinion on that. I think I should leave it up to you guys.

---

In the meantime, I've found a different (non-apache-beam-based) solution to the problem I was working on when I created this pull request. So for me personally, it does not matter anymore, and I'd also be ok with this PR begin closed without further action.

**Comment by RyanSkraba** [comment]
Unfortunately, I think we should close this PR and refine the requirements.  The strictly "right" thing to do would probably be to implement a Beam date type with a finer granularity :/
